mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -742,6 +747,197 @@ extractLayouts(Operation *op, llvm::StringMap<unsigned> &fLayoutMap,
+        for (Value input : genericOp.getInputs()) {
+          FailureOr<bool> maybeHasPointwise =
+              tracesToTarget(input, target, deps, visited);
+          if (succeeded(maybeHasPointwise)) {
+            return true; // Found target through pointwise
+          }
+        }
+      }
+    }
+  }
+
+  return failure();
+}
+
+// Find all reductions and check if they trace back to our GEMM output
+static FusionInfo getFusionInfo(Value gemmResult, GemmFeatures features) {
+  FusionInfo info;
+
+  // Find the target (allocation or block argument)
+  FailureOr<Value> maybeTarget = getBaseValue(gemmResult);
+  if (failed(maybeTarget))
+    return info; // None found
+
+  Value target = *maybeTarget;
+
+  // Get the parent function
+  auto defOp = gemmResult.getDefiningOp();
+  auto funcOp = defOp ? rock::getParentFuncOp(defOp) : nullptr;
+  if (!funcOp) {
+    return info;
+  }
+
+  // Walk all reduce operations and check if they trace back to our GEMM.
+  // Note, we are assuming that all reduce operations are returned here.
+  BufferDependencyAnalysis deps(funcOp);
+  funcOp->walk([&](rock::ReduceOp reduceOp) {
+    DenseSet<Value> visited;
+    FailureOr<bool> maybeHasPointwise =
+        tracesToTarget(reduceOp.getIn(), target, deps, visited);
+
+    if (succeeded(maybeHasPointwise)) {
+      ReductionInfo redInfo;
+      redInfo.method = reduceOp.getReduceMethod();
+      redInfo.axis = reduceOp.getAxis().getSExtValue();
+      auto memrefType = cast<MemRefType>(reduceOp.getIn().getType());
+      redInfo.rank = memrefType.getRank();
+
+      // Extract stride for the reduction dimension
+      SmallVector<int64_t> strides;
+      int64_t offset;
+      if (succeeded(memrefType.getStridesAndOffset(strides, offset))) {
+        redInfo.stride = strides[redInfo.axis];
+      } else {
+        // If we can't determine stride, use dynamic sentinel
+        redInfo.stride = ShapedType::kDynamic;
+      }
+
+      redInfo.hasPointwiseBefore = *maybeHasPointwise;
+      info.reductions.push_back(redInfo);
+    }
+  });
+
+  // Sort reductions for consistent ordering in problem key
+  std::sort(info.reductions.begin(), info.reductions.end());
+
+  return info;
+}
+
+// Append fusion information to the problem key string
+static void appendOutputFusionInfo(llvm::raw_svector_ostream &problemOS,
+                                   const FusionInfo &fusionInfo) {
+  constexpr char sep = ' ';
+
+  if (!fusionInfo.hasReduction())
+    return;
+
+  problemOS << sep << "-fusion_reduce" << sep
+            << "count=" << fusionInfo.numReductionOutputs();
+
+  // Encode each reduction in format: method:rank:axis:stride[:hasPointwise]
+  for (const auto &reduction : fusionInfo.reductions) {
+    problemOS << sep;
+
+    // Add reduction method
+    switch (reduction.method) {
+    case ReduceMethod::Sum:
+      problemOS << "sum";
+      break;
+    case ReduceMethod::Max:
```

Comment:
can there be some other kind of reduce operations ? Something other than Sum/Max ?

i suggest adding `llvm_unreachable` as default case.